[WIP] Hanlin nautobot view set #1812 (#1902)
nautobot/nautobot: nautobot/docs/plugins/development.md

    bulk_update_form_class = YourPluginModelBulkEditForm
    filterset_class = YourPluginModelFilterSet
    filterset_form_class = YourPluginModelFilterForm
    form_class = YourPluginModelForm
    queryset = YourPluginModel.objects.all()
    serializer_class = serializers.YourPluginModelSerializer
    table_class = YourPluginModelTable
```

#### Setting ViewSet Attributes

**One caveat of using the NautobotUIViewSet is that the `queryset`, `serializer_class` and `table_class` attribute of the `YourPluginModelUIViewSet` has to be set before most of the `NautobotUIViewSet` functionalities will become available.**

By default the URL patterns generated by a `NautobotUIViewSet` are based on the model's `slug` (`/model-name/<slug>/` for the detail view, `/model-name/<slug>/edit/` for the edit view, etc.). If your model lacks a `slug` field, or if you otherwise need to use a different field to look up an object, just override the default `lookup_field` in your ViewSet attributes:

```python
from nautobot.core.views.viewsets import NautobotUIViewset

class YourPluginModelUIViewSet(NautobotUIViewSet):
    ...
    lookup_field = "pk"
    ...
```

> ### View Template Context

Reviewer: ```suggestion
#### View Template Context
```

Templates can benefit from a very rich context passed down from the views and renderer, including forms, tables, as well as any other information that may be helpful for rendering templates. The keys it provides are as follows:

* `content_type`: The ContentType object for the associated model
* `filter_form`: The FilterForm object for the associated model
* `form`: A Form object for the associated model if relevant (`None` for list and detail/retrieve views)
* `object`: An instance of the associated mode if available (`None` for list and bulk operation views)
* `permissions`: Summary of user permissions for the given model
* `return_url`: The relevant return URL
* `table`: A Table object for the associated model if relevant (`None` for detail/retrieve and update views)
* `table_config_form`: A TableConfigForm object for the associated `table`, providing the ability to customize the table

> * `verbose_name`: The singular form of the models name
> * `verbose_name_plural`: The plural form of the models name

Reviewer 2: ```suggestion
* `verbose_name`: The singular form of the model's name
* `verbose_name_plural`: The plural form of the model's name
```

An example from editing a Provider object:

```python
{
    'content_type': <ContentType: circuits | provider>,

>     'filter_form': <ProviderFilterForm bound=True, valid=Unknown, fields=(region;site;location;q;asn;tag;cf_emails_circuit_maintenances;cf_provider_parser_circuit_maintenances)>,
>     'form': <ProviderForm bound=False, valid=Unknown, fields=(name;slug;asn;account;portal_url;noc_contact;admin_contact;comments;tags;object_note;cf_emails_circuit_maintenances;cf_provider_parser_circuit_maintenances)>,
>     'object': <Provider: NautobotProivder>,

Reviewer 2: ```suggestion
    'filter_form': <ProviderFilterForm bound=True, valid=Unknown, fields=(region;site;location;q;asn;tag)>,
    'form': <ProviderForm bound=False, valid=Unknown, fields=(name;slug;asn;account;portal_url;noc_contact;admin_contact;comments;tags;object_note)>,
    'object': <Provider: NautobotProvider>,
```

    'permissions': {'add': True, 'change': True, 'delete': True, 'view': True},
    'return_url': '/circuits/providers/nautobotprovider',
    'table': None,
    'table_config_form': None,
    'verbose_name': 'provider',
    'verbose_name_plural': 'providers'
}
```

Other context keys may be available for certain views:

* `editing`: Provided for create and update views to help the template determine if this is a new or existing object
* `action_buttons`: Provided for the list view for the top of table buttons (such as "Add" and "Export")

> You may see other context keys provided by default but these are strictly intended to be transitionary for the time being. Some examples of those are:

Reviewer 2: ```suggestion
You may see other context keys as well, but any not documented above should not be relied upon as they may be removed in a future release. Some examples of those are:
```

class YourPluginModelUIViewSet(
    view_mixins.ObjectListViewMixin,
    view_mixins.ObjectDetailViewMixin,
    view_mixins.ObjectEditViewMixin,
    view_mixins.ObjectDestroyViewMixin,
):

    filterset_class = YourPluginModelFilterSet
    filterset_form_class = YourPluginModelFilterForm
    form_class = YourPluginModelForm
    queryset = YourPluginModel.objects.all()
    serializer_class = serializers.YourPluginModelSerializer
    table_class = YourPluginModelTable
    # You do not need to specify attributes that are not needed.
```

Excluding unwanted urls from `NautobotUIViewSetRouter` is done for you at the ViewSet level. If you do not inherit the unwanted ViewMixins, the corresponding route from the router will not be published.

```python
# urls.py
# All the urls correspond to BulkViewMixins will not be published when you register your ViewSet with the router.
router.register("yourpluginmodel", views.YourPluginModelUIViewSet)
```

#### Template Naming for NautobotUIViewSet

Template naming is very intuitive in NautobotUIViewSet. In `templates/yourpluginmodel` folder, name your templates following this convention `{app_label}/{model_name}_{self.action}.html`.

> | ViewMixins    | self.action   |
> | ------------- |:-------------:|
> | ObjectlistViewMixin        | list         |

Reviewer 2: ```suggestion
| ViewMixins                 | self.action  |
| -------------------------- |:------------:|
| ObjectListViewMixin        | list         |
```

| ObjectDetailViewMixin      | retrieve     |
| ObjectEditViewMixin        | create/update|
| ObjectDestroyViewMixin     | destroy      |
| ObjectBulkDestroyViewMixin | bulk_destroy |
| ObjectBulkCreateViewMixin  | bulk_create  |
| ObjectBulkUpdateViewMixin  | bulk_update  |

> For example, for a DetailView template for `yourpluginmodel`, the template name will be `yourplugin/yourpluginmodel_retrieve.html`, for a BulkCreateView template for `yourpluginmodel`, the template name will be `yourplugin/yourpluginmodel_bulk_create.html` and etc.
> 
> If you do not specify your own templates in `templates/yourpluginmodel` folder, `NautobotUIViewSet` will fall back to `generic/object_{self.action}.html`.

Reviewer 2: ```suggestion
For example, for a DetailView template for `YourPluginModel`, the template name will be `yourplugin/yourpluginmodel_retrieve.html`, for a BulkCreateView template for `yourpluginmodel`, the template name will be `yourplugin/yourpluginmodel_bulk_create.html` and etc.

If you do not provide your own templates in the `yourplugin/templates/yourplugin` folder, `NautobotUIViewSet` will fall back to `generic/object_{self.action}.html`.
```

except ImportError:
    pass
```

> ## NautobotUIViewSet

Reviewer 2: Should this be moved up so that it falls under the `Adding Web UI Views` section heading?

Note that this ViewSet is catered specifically to the UI, not the API.

Concrete examples on how to use `NautobotUIViewSet` resides in `nautobot.circuits.views`.

Below we provide an example on how to use `NautobotUIViewSet` on a theoretical plugin model.

```python
from nautobot.core.views.viewsets import NautobotUIViewset

class YourPluginModelUIViewSet(NautobotUIViewSet):
    bulk_create_form_class = YourPluginModelCSVForm
    bulk_update_form_class = YourPluginModelBulkEditForm
    filterset_class = YourPluginModelFilterSet
    filterset_form_class = YourPluginModelFilterForm
    form_class = YourPluginModelForm
    queryset = YourPluginModel.objects.all()
    serializer_class = serializers.YourPluginModelSerializer
    table_class = YourPluginModelTable
```

### Setting ViewSet Attributes

> **One caveat of using the NautobotUIViewSet is that the `queryset`, `serializer_class` and `table_class` attribute of the `YourPluginModelUIViewSet` has to be set before most of the `NautobotUIViewSet` functionalities will become available.**

Reviewer 2: Thinking about this, the `serializer_class` attribute would be nice to not require in the future, so that it's possible to use NautobotUIViewSet before implementing the REST API for a model.

> `NautobotUIViewSet`'s `lookup_field` is default to `slug` as we transition from `pk` to `slug` across Nautobot Apps. If you want to use a different field to look up an object, just override the default `lookup_field` in your ViewSet attributes:

Reviewer 2: ```suggestion
By default the URL patterns generated by a `NautobotUIViewSet` are based on the model's `slug` (`/model-name/<slug>/` for the detail view, `/model-name/<slug>/edit/` for the edit view, etc.). If your model lacks a `slug` field, or if you otherwise need to use a different field to look up an object, just override the default `lookup_field` in your ViewSet attributes:
```

    table_class = YourPluginModelTable
    # You do not need to specify attributes that are not needed.
```

### Template Naming for NautobotUIViewSet

Template naming is very intuitive in NautobotUIViewSet. In `templates/yourpluginmodel` folder, name your templates following this convention `{app_label}/{model_name}_{self.action}.html`.

| ViewMixins    | self.action   |
| ------------- |:-------------:|
| ObjectlistViewMixin        | list         |
| ObjectDetailViewMixin      | retrieve     |
| ObjectEditViewMixin        | create/update|
| ObjectDestroyViewMixin     | destroy      |
| ObjectBulkDestroyViewMixin | bulk_destroy |
| ObjectBulkCreateViewMixin  | bulk_create  |
| ObjectBulkUpdateViewMixin  | bulk_update  |

For example, for a DetailView template for `yourpluginmodel`, the template name will be `yourplugin/yourpluginmodel_retrieve.html`, for a BulkCreateView template for `yourpluginmodel`, the template name will be `yourplugin/yourpluginmodel_bulk_create.html` and etc.

If you do not specify your own templates in `templates/yourpluginmodel` folder, `NautobotUIViewSet` will fall back to `generic/object_{self.action}.html`.

> If you do specify a `{app_label}/{model_opts.model_name}_create.html` file but not a `{app_label}/{model_opts.model_name}_update.html` file. When you update an object, it will fall back to `{app_label}/{model_opts.model_name}_create.html` and vice versa.

Reviewer 2: ```suggestion
Since in many cases the `create` and `update` templates for a model will be identical, you are not required to create both. If you provide a `{app_label}/{model_opts.model_name}_create.html` file but not a `{app_label}/{model_opts.model_name}_update.html` file, then when you update an object, it will fall back to `{app_label}/{model_opts.model_name}_create.html` and vice versa.
```

## NautobotUIViewSetRouter

With `NautobotUIViewSet` as the base UI ViewSet for `YourPluginModel`, it is required to register your urls with the help of `NautobotUIViewSetRouter`.

For a concrete example on how to use `NautobotUIViewSetRouter`, see `nautobot.circuits.urls`.

Below is a theoretical `urls.py` file for `YourPluginModel`:

```python
from django.urls import path

from nautobot.core.views.routers import NautobotUIViewSetRouter
from nautobot.plugins import views

router = NautobotUIViewSetRouter()
router.register("yourpluginmodel", views.YourPluginModelUIViewSet)

urlpatterns = [
    # Extra urls that do not follow the patterns of `NautobotUIViewSetRouter` go here.
    # changelog, notes and etc.
    ...
    path(
        "yourpluginmodels/<slug:slug>/changelog/",
        ObjectChangeLogView.as_view(),
        name="yourpluginmodel_changelog",
        kwargs={"model": yourpluginmodel},
    ),
    path(
        "yourpluginmodels/<slug:slug>/notes/",
        ObjectNotesView.as_view(),
        name="yourpluginmodel_notes",
        kwargs={"model": yourpluginmodel},
    ),
    ...
]
urlpatterns += router.urls
```

> ### Excluding Urls from NautobotUIViewSetRouter

Reviewer 2: Should this be a separate section or can it just be an additional sentence or two under `Excluding ViewMixins from NautobotUIViewSet`?

Author: I can move it under `Excluding ViewMixins from NautobotUIViewSet`
